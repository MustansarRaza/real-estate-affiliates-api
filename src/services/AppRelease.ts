import { AppRelease } from "affiliates-api/models";
import { Op } from "sequelize";

export interface AppVariablesDTO {
  min_version: string;
}
export interface SettingsDTO {
  version: string;
  appVariables: AppVariablesDTO;
}
class AppReleaseService {
  /**
   * Returns mobile application settings
   */
  async get(appName: string): Promise<SettingsDTO | null> {
    const release = await AppRelease.findOne({
      attributes: ["version", "appVariables"],
      where: {
        appName: {
          [Op.eq]: appName,
        },
      },
    });

    return release ? this.serialize(release) : null;
  }
  /**
   * Serializes a queried {@see AppRelease} model instance into a DTO.
   */
  private serialize(release: AppRelease): SettingsDTO {
    return {
      version: release.version.toString(),
      appVariables: {
        min_version: release.appVariables!.min_version,
      },
    };
  }
}
export default AppReleaseService;
