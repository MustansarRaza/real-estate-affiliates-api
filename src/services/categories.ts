import { ListingStatusKey, PlatformKey } from "@pf/taxonomies";
import {
  Address,
  AgencyConfiguration,
  Category,
  Listing,
  ListingStatus,
  Location,
  Platform,
  popularCategoryKeyToIndex,
  Project,
  ProjectAffiliateVisibilityID,
  ProjectStatus,
} from "affiliates-api/models";
import { col, fn, Op } from "sequelize";

interface CategoryAvailabilityDTO {
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

interface CategoryCityAvailabilityDTO {
  cityID: number;
  unitsCount: number;
}

interface CategoryDTO {
  id: number;
  name: string;
  nameSingular: string;
  orderIndex: number;
  slug: string;
  parentID: number | null;
  popularityIndex: number;
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

/**
 * Provides access to categories.
 */
class CategoriesService {
  /**
   * Gets a list of all available categories.
   */
  async list(agencyID: number | null): Promise<CategoryDTO[]> {
    let platformId = null;

    if (agencyID) {
      const agencyConfig = await AgencyConfiguration.findOne({
        attributes: ["platformId"],
        where: { agencyId: { [Op.eq]: agencyID } },
        raw: true,
      });
      platformId = agencyConfig?.platformId;
    }
    if (!platformId) {
      const platform = await Platform.findOne({
        attributes: ["id"],
        where: { platformKey: { [Op.eq]: PlatformKey.ZAMEEN } },
        raw: true,
      });
      platformId = platform.id;
    }

    const categories = await Category.findAll({
      where: { platformId: { [Op.eq]: platformId } },
      order: [
        ["type_order", "ASC"],
        ["type_id", "ASC"],
      ],
    });

    const categoryCitiesAndProjectsMap = await this.getCategoryCitiesAndProjectsMap(agencyID);

    return categories.map((category: Category) => ({
      id: category.id,
      name: category.name,
      nameSingular: category.nameSingular,
      orderIndex: category.orderIndex,
      slug: category.slug,
      parentID: category.parentID,
      popularityIndex: popularCategoryKeyToIndex(category.key),
      availableInCities: categoryCitiesAndProjectsMap[category.id]?.availableInCities || [],
      availableInProjects: categoryCitiesAndProjectsMap[category.id]?.availableInProjects || [],
    }));
  }

  /**
   * Gets a map of cities and Projects against a category in which
   * that category is available.
   */
  private async getCategoryCitiesAndProjectsMap(
    agencyID: number | null
  ): Promise<{
    [propertTypeId: number]: CategoryAvailabilityDTO;
  }> {
    const categoriesAvailablityResult = await Listing.findAll({
      attributes: [[fn("COUNT", "id"), "Count"]],
      include: [
        {
          model: ListingStatus,
          attributes: [],
          where: {
            key: {
              [Op.eq]: ListingStatusKey.AVAILABLE,
            },
          },
          required: true,
        },
        {
          model: Category,
          attributes: ["id"],
          required: true,
        },
        {
          model: Project,
          required: true,
          attributes: ["id"],
          where: {
            status: {
              [Op.eq]: ProjectStatus.ON,
            },
            affiliateVisibility: {
              [Op.eq]: ProjectAffiliateVisibilityID.VISIBLE,
            },
            ...(agencyID && {
              agencyID: {
                [Op.eq]: agencyID,
              },
            }),
          },
          include: [
            {
              model: Address,
              required: true,
              attributes: [],
              include: [
                {
                  model: Location,
                  attributes: ["id"],
                  required: true,
                },
              ],
            },
          ],
        },
      ],
      group: [
        col("Category.type_id"),
        col("Project.Address.Location.title"),
        col("Project.project_id"),
      ],
      raw: true,
    });

    const categoryCitiesAndProjectsMap: { [propertTypeId: number]: CategoryAvailabilityDTO } = {};
    categoriesAvailablityResult.forEach((row: any) => {
      const propertyTypeId = row["Category.id"];
      const projectID = row["Project.id"];
      const cityID = row["Project.Address.Location.id"];
      const unitsCount = row["Count"];

      if (categoryCitiesAndProjectsMap.hasOwnProperty(propertyTypeId)) {
        categoryCitiesAndProjectsMap[propertyTypeId].availableInProjects.push(projectID);
        const index = categoryCitiesAndProjectsMap[propertyTypeId].availableInCities
          .map((city: CategoryCityAvailabilityDTO) => city.cityID)
          .indexOf(cityID);
        if (index === -1) {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities.push({
            cityID,
            unitsCount,
          });
        } else {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities[
            index
          ].unitsCount += unitsCount;
        }
      } else {
        categoryCitiesAndProjectsMap[propertyTypeId] = {
          availableInProjects: [projectID],
          availableInCities: [{ cityID, unitsCount }],
        };
      }
    });

    return categoryCitiesAndProjectsMap;
  }
}

export default CategoriesService;
