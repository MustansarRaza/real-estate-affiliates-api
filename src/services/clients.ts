import { SourcePlatform } from "@pf/taxonomies";
import { AppError } from "@pf/utils";
import {
  AffiliateContact,
  Client,
  ClientPhoneNumber,
  ClientReferralType,
  ClientStatusID,
} from "affiliates-api/models";
import { ISO8601DateTime } from "affiliates-api/types";
import { parsePhoneNumberFromString, PhoneNumber } from "libphonenumber-js";
import { Op, Sequelize, WhereOptions } from "sequelize";
import LocationsService from "./locations";
import MeService from "./me";

export interface ClientPhoneNumberDTO {
  id: number;
  value: string;
}

export interface ClientDTO {
  id: number;
  name: string;
  email: string | null;
  phoneNumbers: ClientPhoneNumberDTO[];
  address: string | null;
  locationID: number | null;
  comment: string | null;
  createdAt: ISO8601DateTime;
  updatedAt: ISO8601DateTime;
}

export interface CreateClientPhoneNumberDTO {
  value: string;
}

export interface CreateClientDTO {
  name: string;
  email: string;
  phoneNumbers: CreateClientPhoneNumberDTO[];
  address: string | null;
  locationID: number | null;
  comment: string | null;
  source?: SourcePlatform;
}

interface OtherServices {
  me: MeService;
  locations: LocationsService;
}

interface BySSOIDLookup {
  ssoID: string;
}

type ByNamePhoneNumberOrEmailLookup = RequireAtLeastOne<{
  name?: string | null;
  phoneNumber?: string | null;
  email?: string | null;
}>;

/**
 * Provides access to clients.
 *
 * Clients are essentially contacts. They represent a person which can
 * be assigned to one or more leads.
 */
class ClientsService {
  constructor(private services: OtherServices) {}

  /**
   * Gets a flat list of clients for the specified potential
   * affiliate.
   */
  async list({ ssoID }: BySSOIDLookup): Promise<ClientDTO[]> {
    const clients = await Client.findAll({
      where: {
        statusID: {
          [Op.eq]: ClientStatusID.ACTIVE,
        },
        referralType: {
          [Op.eq]: ClientReferralType.AFFILIATE,
        },
      },
      include: [
        {
          model: AffiliateContact,
          required: true,
          attributes: [],
          where: {
            ssoID: {
              [Op.eq]: ssoID,
            },
          },
        },
        {
          model: ClientPhoneNumber,
          separate: true,
          order: [["cc_id", "ASC"]],
        },
      ],
      order: [
        ["time_added", "DESC"],
        ["clientid", "DESC"],
      ],
    });

    return clients.map((client: Client) => this.serialize(client));
  }

  /**
   * Gets an existing client by name, phone number or email.
   *
   * This is used to stop users from creating a client that another user
   * already has. This is _not_ to be used to avoid inserting duplicate
   * records. This is NOT CONCURRENCY SAFE.
   */
  async find({ email, phoneNumber }: ByNamePhoneNumberOrEmailLookup): Promise<ClientDTO | null> {
    const filters: WhereOptions[] = [];
    const phoneNumberFilters: WhereOptions[] = [];

    // e-mail addresses are not case sensitive so we lower-case on both
    // ends to simulate a case-insensitive search.. we can't use ILIKE
    // because its not supported by older MySQL versions or SQLite
    if (email) {
      const sanitizedEmail = email.toLowerCase().trim();
      filters.push(Sequelize.where(Sequelize.fn("lower", Sequelize.col("email")), sanitizedEmail));
    }

    if (phoneNumber) {
      const sanitizedPhoneNumber = this.sanitizePhoneNumber(phoneNumber);

      phoneNumberFilters.push({
        value: {
          [Op.eq]: sanitizedPhoneNumber,
        },
      });
    }

    const phoneNumbers = await ClientPhoneNumber.findAll({
      where: {
        [Op.or]: phoneNumberFilters,
      },
    });

    if (phoneNumbers) {
      filters.push({
        id: {
          [Op.in]: phoneNumbers.map((entry: ClientPhoneNumber) => entry.clientID),
        },
      });
    }

    const client = await Client.findOne({
      where: {
        [Op.or]: filters,
      },
      include: [
        {
          model: ClientPhoneNumber,
          order: [["cc_id", "ASC"]],
          separate: true,
        },
      ],
    });

    if (!client) {
      return null;
    }

    return this.serialize(client);
  }

  /**
   * Creates a new client that is owned by the specified {@paramref owner}.
   *
   * This will only work if the owner is a verified affiliate and will
   * fail with {@see AppError} if its not.
   */
  async create({ ssoID }: BySSOIDLookup, dto: CreateClientDTO): Promise<ClientDTO> {
    const { name, email, address, locationID, comment, phoneNumbers, source } = dto;

    const [owner, countryID] = await Promise.all([
      this.services.me.owner({ ssoID }),
      // lookup countryID because countries are included
      // in the location tree, we don't expose countryID to
      // the outside world, but we need to set it
      // for backwards compatibility reasons
      locationID ? this.services.locations.countryID({ locationID }) : Promise.resolve(null),
    ]);

    if (!owner) {
      throw new AppError("Current user has no owner", {
        statusCode: 400,
        extra: {
          ssoID,
        },
      });
    }

    if (locationID && !countryID) {
      throw new AppError("Cannot find location or location's country", {
        statusCode: 400,
        extra: {
          locationID,
        },
      });
    }

    // setting this for backwards compatibility, phone numbers
    // are stored separately as a list in zn_pf_client_cells
    const phoneNumber = dto.phoneNumbers[0]?.value || null;

    // make sure the user does not take ownership of a client
    // that somebody else already entered into the database
    // this is a business requirement, not a technical solution
    // to preventing duplicates (we have unique constraints for that)
    const existingClient = await this.find({ email, phoneNumber });
    if (existingClient) {
      throw new AppError("Client already exists", {
        statusCode: 409,
        extra: {
          name,
          email,
          phoneNumber,
        },
      });
    }

    const client = await Client.create(
      {
        statusID: ClientStatusID.ACTIVE,
        source: source || SourcePlatform.AFFILIATES_MOBILE_APP,
        userID: owner.userID,
        agencyID: owner.agencyID,
        name,
        email,
        address,
        countryID,
        locationID,
        comment,
        referralType: ClientReferralType.AFFILIATE,
        referralID: owner.affiliateID,
        ClientPhoneNumbers: phoneNumbers.map((phoneNumber: CreateClientPhoneNumberDTO) => ({
          value: phoneNumber.value,
          // add the short version too because the staff app duplication check is based on the short version
          shortValue: phoneNumber.value.slice(-9),
        })),
      },
      {
        include: [ClientPhoneNumber],
      }
    );

    return this.serialize(client);
  }

  /**
   * Serializes a queried {@see Client} model instance into a DTO.
   */
  private serialize(client: Client): ClientDTO {
    return {
      id: client.id,
      name: client.name,
      email: client.email || null,
      phoneNumbers: client.ClientPhoneNumbers.map((phoneNumber: ClientPhoneNumber) => ({
        id: phoneNumber.id,
        value: phoneNumber.value,
      })),
      address: client.address || null,
      locationID: client.locationID || null,
      comment: client.comment || null,
      createdAt: client.createdAt.toISOString(),
      updatedAt: client.updatedAt.toISOString(),
    };
  }

  /**
   * Sanitizes the specified phone number into a valid E.164 phone number.
   */
  private sanitizePhoneNumber(value: string): string {
    const errorMessage = `${value} is not a valid phone number`;
    const errorData = { statusCode: 400 };

    let parsedPhoneNumber: PhoneNumber | undefined | null = null;

    try {
      parsedPhoneNumber = parsePhoneNumberFromString(value);
    } catch (error) {
      throw new AppError(errorMessage, { ...errorData, error });
    }

    if (!parsedPhoneNumber) {
      throw new AppError(errorMessage, errorData);
    }

    return parsedPhoneNumber.number as string;
  }
}

export default ClientsService;
