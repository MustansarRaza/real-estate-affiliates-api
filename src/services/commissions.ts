import { AffiliateCommissionType } from "@pf/taxonomies";
import { AffiliateCommission } from "affiliates-api/models";
import { maxBy, minBy } from "lodash";
import { Op, WhereOptions } from "sequelize";

export interface ProjectCommissionDTO {
  max: number | null;
  min: number | null;
}

export interface AffiliateCommissionDTO {
  [commissionType: string]: {
    [propertyTypeId: number]: AffiliateCommission[];
  };
}

export interface AffiliateProjectCommissionDTO {
  [projectID: number]: AffiliateCommissionDTO;
}

type ByProjectIDLookup = {
  projectIDs: number[];
};

type ByAffiliateTypeIDLookup = {
  affiliateTypeId: number | null;
};

type ByFiltersLookup = {
  createdBy?: number;
  propertyTypeId?: number;
  commissionType?: AffiliateCommissionType;
};

type AffiliateCommissionLookupOptions = ByProjectIDLookup &
  ByAffiliateTypeIDLookup &
  ByFiltersLookup;

class CommissionsService {
  async getDetailsFor(
    options: AffiliateCommissionLookupOptions
  ): Promise<{ [projectID: number]: AffiliateCommission[] }> {
    const unformattedDate = new Date();
    unformattedDate.setHours(0, 0, 0, 0);
    const currentDate = unformattedDate.toISOString();

    const commissionDetails = await AffiliateCommission.findAll({
      where: {
        [Op.and]: this.createFiltersForAffiliateCommission(options),
        [Op.or]: [
          {
            startDate: {
              [Op.is]: null,
            },
            endDate: {
              [Op.is]: null,
            },
          },
          {
            startDate: {
              [Op.not]: null,
              [Op.lte]: currentDate,
            },
            endDate: {
              [Op.is]: null,
            },
          },
          {
            startDate: {
              [Op.is]: null,
            },
            endDate: {
              [Op.not]: null,
              [Op.gte]: currentDate,
            },
          },
          {
            startDate: {
              [Op.not]: null,
              [Op.lte]: currentDate,
            },
            endDate: {
              [Op.not]: null,
              [Op.gte]: currentDate,
            },
          },
        ],
      },
      raw: true,
    });

    const projectAndCommissionsMapping: { [projectID: number]: AffiliateCommission[] } = {};
    commissionDetails.forEach((row: AffiliateCommission) => {
      if (projectAndCommissionsMapping.hasOwnProperty(row.projectId))
        projectAndCommissionsMapping[row.projectId].push(row);
      else projectAndCommissionsMapping[row.projectId] = [row];
    });

    return projectAndCommissionsMapping;
  }

  async getCommissionsBreakdownForAffiliate(
    options: AffiliateCommissionLookupOptions
  ): Promise<AffiliateProjectCommissionDTO> {
    const projectAndCommissionsMapping = await this.getDetailsFor(options);

    const avaiableCommissionProjectIDs = Object.keys(projectAndCommissionsMapping);

    const affiliateCommissionBreakdown: AffiliateProjectCommissionDTO = {};
    avaiableCommissionProjectIDs.forEach((projectID) => {
      const projectCommissions = projectAndCommissionsMapping[parseInt(projectID, 10)] || [];

      projectCommissions.forEach((commission) => {
        if (affiliateCommissionBreakdown.hasOwnProperty(commission.projectId))
          if (
            affiliateCommissionBreakdown[commission.projectId].hasOwnProperty(
              commission.commissionType
            )
          )
            if (
              affiliateCommissionBreakdown[commission.projectId][
                commission.commissionType
              ].hasOwnProperty(commission.propertyTypeId)
            )
              affiliateCommissionBreakdown[commission.projectId][commission.commissionType][
                commission.propertyTypeId
              ].push(commission);
            else
              affiliateCommissionBreakdown[commission.projectId][commission.commissionType][
                commission.propertyTypeId
              ] = [commission];
          else
            affiliateCommissionBreakdown[commission.projectId][commission.commissionType] = {
              [commission.propertyTypeId]: [commission],
            };
        else
          affiliateCommissionBreakdown[commission.projectId] = {
            [commission.commissionType]: { [commission.propertyTypeId]: [commission] },
          };
      });
    });

    return affiliateCommissionBreakdown;
  }

  async getProjectCommissionStats(
    options: AffiliateCommissionLookupOptions
  ): Promise<Map<number, ProjectCommissionDTO>> {
    const projectAndCommissionsMapping = await this.getDetailsFor(options);

    const projectCommissionStats = new Map<number, ProjectCommissionDTO>();
    Object.keys(projectAndCommissionsMapping).forEach((projectID) => {
      const maxCommission = maxBy(
        projectAndCommissionsMapping[+projectID],
        (affiliateCommission) =>
          affiliateCommission.commission + affiliateCommission.classifiedCredits!
      );
      const minCommission = minBy(
        projectAndCommissionsMapping[+projectID],
        (affiliateCommission) =>
          affiliateCommission.commission + affiliateCommission.classifiedCredits!
      );

      return projectCommissionStats.set(+projectID, {
        max: maxCommission?.commission! + maxCommission?.classifiedCredits!,
        min: minCommission?.commission! + minCommission?.classifiedCredits!,
      });
    });

    return projectCommissionStats;
  }

  private createFiltersForAffiliateCommission(
    options: AffiliateCommissionLookupOptions
  ): WhereOptions[] {
    const filters: WhereOptions[] = [
      {
        affiliateTypeId: {
          [Op.eq]: options.affiliateTypeId,
        },
        projectId: {
          [Op.in]: options.projectIDs,
        },
      },
    ];

    if (options?.createdBy) {
      filters.push({
        createdBy: {
          [Op.eq]: options.createdBy,
        },
      });
    }
    if (options?.propertyTypeId) {
      filters.push({
        propertyTypeId: {
          [Op.eq]: options.propertyTypeId,
        },
      });
    }
    if (options?.commissionType) {
      filters.push({
        commissionType: {
          [Op.eq]: options.commissionType,
        },
      });
    }

    return filters;
  }
}

export default CommissionsService;
