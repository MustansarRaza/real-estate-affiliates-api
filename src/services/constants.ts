export enum ErrorMessages {
  USER_AFFILIATE_ERROR_MESSAGE = "Current user is not an affiliate",
}

export const FORMAT_DATE = {
  FULL_DATE_24_HOUR_FORMAT_DATE_FNS: "YYYY-MM-DD HH:mm:ss",
};
