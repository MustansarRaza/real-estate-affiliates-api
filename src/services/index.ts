import AppRelease from "./AppRelease";
import CategoriesService from "./categories";
import ClientsService, { CreateClientDTO as CreateClientDTOType } from "./clients";
import CommissionsService from "./commissions";
import LeadsService, { CreateLeadDTO as CreateLeadDTOType } from "./leads";
import ListingsService, { ListingDTO as ListingDTOType } from "./listings";
import LocationsService from "./locations";
import MeService, { CreateMeDTO as CreateMeDTOType, MeDTO as MeDTOType } from "./me";
import ProfileService from "./profile";
import ProjectsService from "./projects";
import ListingReservationService, { ListingReservationOptions } from "./reservations";
import TasksService from "./tasks";
import TaskTypesService from "./taskTypes";

export interface ServiceMesh {
  profile: ProfileService;
  me: MeService;
  clients: ClientsService;
  tasks: TasksService;
  taskTypes: TaskTypesService;
  leads: LeadsService;
  projects: ProjectsService;
  listings: ListingsService;
  locations: LocationsService;
  categories: CategoriesService;
  reservations: ListingReservationService;
  commissions: CommissionsService;
  appRelease: AppRelease;
}

const useServices = (): ServiceMesh => {
  const locations = new LocationsService();
  const profile = new ProfileService({ locations });
  const me = new MeService({ profile, locations });
  const clients = new ClientsService({ me, locations });
  const taskTypes = new TaskTypesService();
  const leads = new LeadsService({ me, taskTypes });
  const tasks = new TasksService({ leads, taskTypes });
  const listings = new ListingsService({ me });
  const commissions = new CommissionsService();
  const projects = new ProjectsService({ me, listings, commissions, locations });
  const categories = new CategoriesService();
  const reservations = new ListingReservationService({ me, listings });
  const appRelease = new AppRelease();

  return {
    profile,
    me,
    clients,
    leads,
    tasks,
    taskTypes,
    projects,
    listings,
    locations,
    categories,
    reservations,
    commissions,
    appRelease,
  };
};

export {
  useServices,
  ClientsService,
  LeadsService,
  MeService,
  ProfileService,
  ProjectsService,
  ListingsService,
  TasksService,
  TaskTypesService,
  LocationsService,
  CategoriesService,
  ListingReservationService,
  CommissionsService,
  AppRelease,
};

export type CreateClientDTO = CreateClientDTOType;
export type CreateLeadDTO = CreateLeadDTOType;
export type CreateMeDTO = CreateMeDTOType;
export type CreateReservationDTO = ListingReservationOptions;
export type ListingDTO = ListingDTOType;
export type MeDTO = MeDTOType;
