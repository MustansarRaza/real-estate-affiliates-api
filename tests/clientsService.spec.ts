import { SourcePlatform } from "@pf/taxonomies";
import {
  Client,
  ClientAccessID,
  ClientPhoneNumber,
  ClientReferralType,
  ClientStatusID,
  Country,
  Location,
} from "affiliates-api/models";
import { ClientsService, LocationsService } from "affiliates-api/services";
import { Op } from "sequelize";
import { loadSequelizeFixtures, useMockedServices } from "./support";
import {
  createClient,
  createClientPhoneNumber,
  ModelFixtureConstants,
  ModelFixtureNames,
} from "./support/data";

describe("ClientsService", () => {
  const mockAgencyID = 1337;

  const mockedServices = useMockedServices();
  const service = new ClientsService({ me: mockedServices.me, locations: new LocationsService() });

  beforeEach(async () => {
    jest.resetAllMocks();

    mockedServices.me.owner.mockReturnValue(
      Promise.resolve({
        affiliateID: ModelFixtureConstants.verifiedAffiliateID,
        userID: ModelFixtureConstants.staffUserBDMID,
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      })
    );

    await loadSequelizeFixtures(
      ModelFixtureNames.AGENTS,
      ModelFixtureNames.CONTACTS,
      ModelFixtureNames.AFFILIATES,
      ModelFixtureNames.AFFILIATE_CONTACTS
    );
  });

  describe("list", () => {
    it("only lists active clients", async () => {
      const activeClient = await createClient({
        referralID: ModelFixtureConstants.verifiedAffiliateID,
        statusID: ClientStatusID.ACTIVE,
      });
      await createClient({ statusID: ClientStatusID.DELETED });

      const clients = await service.list({ ssoID: ModelFixtureConstants.verifiedSSOID });

      expect(clients).toHaveLength(1);
      expect(clients[0].id).toBe(activeClient.id);
    });

    it("only lists clients belonging to affiliates", async () => {
      const affiliateClient = await createClient({
        referralID: ModelFixtureConstants.verifiedAffiliateID,
        referralType: ClientReferralType.AFFILIATE,
      });
      await createClient({
        referralID: ModelFixtureConstants.verifiedAffiliateID,
        referralType: ClientReferralType.DIRECT,
      });

      const clients = await service.list({ ssoID: ModelFixtureConstants.verifiedSSOID });

      expect(clients).toHaveLength(1);
      expect(clients[0].id).toBe(affiliateClient.id);
    });

    it("only lists clients belonging to the specified potential affiliate", async () => {
      const myClient = await createClient({
        referralID: ModelFixtureConstants.verifiedAffiliateID,
      });
      await createClient({ referralID: 33 });

      const clients = await service.list({ ssoID: ModelFixtureConstants.verifiedSSOID });

      expect(clients).toHaveLength(1);
      expect(clients[0].id).toBe(myClient.id);
    });

    it("lists phone numbers of the client", async () => {
      const client = await createClient({ referralID: ModelFixtureConstants.verifiedAffiliateID });

      const phoneNumber1 = await ClientPhoneNumber.create({
        clientID: client.id,
        value: "+923212473743",
        shortValue: "212473743",
      });

      const phoneNumber2 = await ClientPhoneNumber.create({
        clientID: client.id,
        value: "+923006666025",
        shortValue: "006666025",
      });

      const clients = await service.list({ ssoID: ModelFixtureConstants.verifiedSSOID });
      expect(clients).toHaveLength(1);
      expect(clients[0].phoneNumbers).toHaveLength(2);

      expect(clients[0].phoneNumbers[0]).toStrictEqual({
        id: phoneNumber1.id,
        value: phoneNumber1.value,
      });

      expect(clients[0].phoneNumbers[1]).toStrictEqual({
        id: phoneNumber2.id,
        value: phoneNumber2.value,
      });
    });
  });

  describe("find", () => {
    it("can find by email and is not case sensitive", async () => {
      const client = await createClient({ email: "tEsT@test.COM" });

      const result = await service.find({ email: "test@tesT.com" });
      expect(result?.id).toBe(client.id);
    });

    it("can find by an E.164 formatted phone number", async () => {
      const client = await createClient();
      await createClientPhoneNumber(client.id, { value: "+923005323674" });

      const result = await service.find({ phoneNumber: "+923005323674" });
      expect(result?.id).toBe(client.id);
    });

    it("raises an error if an invalid number is specified", async () => {
      await expect(service.find({ phoneNumber: "ab" })).rejects.toMatchSnapshot();
    });
  });

  describe("create", () => {
    const createClientBaseDTO = {
      name: "Walt Disney",
      email: "walt@disney.com",
      phoneNumbers: [
        {
          value: "+923214444444",
        },
      ],
      address: null,
      locationID: null,
      comment: "this is me",
      source: SourcePlatform.AFFILIATES_MOBILE_APP,
    };

    it("creates a new client and links it to the right affiliate", async () => {
      const clientDTO = await service.create(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        createClientBaseDTO
      );

      const client: Client | null = await Client.findOne({
        where: {
          id: {
            [Op.eq]: clientDTO.id,
          },
        },
      });

      expect(client).toBeTruthy();
      expect(client?.id).toBe(clientDTO.id);
      expect(client?.statusID).toBe(ClientStatusID.ACTIVE);
      expect(client?.source).toBe(SourcePlatform.AFFILIATES_MOBILE_APP);
      expect(client?.typeID).toBe(null);
      expect(client?.accessID).toBe(ClientAccessID.UNKNOWN);
      expect(client?.agencyID).toBe(mockAgencyID);
      expect(client?.userID).toBe(ModelFixtureConstants.staffUserBDMID);
      expect(client?.name).toBe(clientDTO.name);
      expect(client?.email).toBe(clientDTO.email);
      expect(client?.address).toBe(clientDTO.address);
      expect(client?.locationID).toBe(clientDTO.locationID);
      expect(client?.countryID).toBe(null);
      expect(client?.phoneNumber).toBe(null);
      expect(client?.comment).toBe(clientDTO.comment);
      expect(client?.referralType).toBe(ClientReferralType.AFFILIATE);
      expect(client?.referralID).toBe(ModelFixtureConstants.verifiedAffiliateID);
      expect(client?.createdAt).toBeTruthy();
      expect(client?.updatedAt).toBeTruthy();
    });

    it("creates a ClientPhoneNumber as expected", async () => {
      const clientDTO = await service.create(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        createClientBaseDTO
      );

      const phoneNumber = await ClientPhoneNumber.findOne({
        where: {
          clientID: {
            [Op.eq]: clientDTO.id,
          },
        },
      });

      expect(phoneNumber).toBeTruthy();
      expect(phoneNumber?.value).toBe(clientDTO.phoneNumbers[0].value);
      expect(phoneNumber?.shortValue).toBe(clientDTO.phoneNumbers[0].value.slice(-9));
    });

    it("looks up the country ID for the specified location ID and stores it", async () => {
      const country = await Country.create({
        name: "Pakistan",
      });

      const location = await Location.create({
        name: "Lahore",
        level: 1,
        countryID: country.id,
      });

      const clientDTO = await service.create(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        {
          ...createClientBaseDTO,
          locationID: location.id,
        }
      );

      const client: Client | null = await Client.findOne({
        where: {
          id: {
            [Op.eq]: clientDTO.id,
          },
        },
      });

      expect(client?.countryID).toBe(country.id);
      expect(client?.locationID).toBe(location.id);
    });

    it("does not allow creating clients without having an owner", async () => {
      mockedServices.me.owner.mockReturnValue(Promise.resolve(null));

      await expect(
        service.create({ ssoID: ModelFixtureConstants.verifiedSSOID }, createClientBaseDTO)
      ).rejects.toMatchSnapshot();
    });

    it("allows creating clients with the same name", async () => {
      const client1 = await createClient({ name: "Walt Disney" });

      const client2 = await service.create(
        { ssoID: ModelFixtureConstants.verifiedSSOID },
        {
          ...createClientBaseDTO,
          name: "Walt Disney",
        }
      );

      expect(client1.id).not.toBe(client2.id);
    });

    it("does not allow creating clients with the same cell number", async () => {
      const client = await createClient();
      await createClientPhoneNumber(client.id, { value: "+923214444444" });

      await expect(
        service.create(
          { ssoID: ModelFixtureConstants.verifiedSSOID },
          {
            ...createClientBaseDTO,
            phoneNumbers: [
              {
                value: "+923214444444",
              },
            ],
          }
        )
      ).rejects.toMatchSnapshot();
    });

    it("does not allow creating clients with the same email address", async () => {
      await createClient({ email: "cookies@monster.com" });

      await expect(
        service.create(
          { ssoID: ModelFixtureConstants.verifiedSSOID },
          {
            ...createClientBaseDTO,
            email: "cookies@monster.com",
          }
        )
      ).rejects.toMatchSnapshot();
    });

    it("does not allow creating clients with an invalid location ID", async () => {
      await expect(
        service.create(
          { ssoID: ModelFixtureConstants.verifiedSSOID },
          {
            ...createClientBaseDTO,
            locationID: 9999,
          }
        )
      ).rejects.toMatchSnapshot();
    });
  });
});
