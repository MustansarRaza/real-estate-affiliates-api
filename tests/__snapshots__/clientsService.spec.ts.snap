// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientsService create does not allow creating clients with an invalid location ID 1`] = `
Object {
  "extra": Object {
    "locationID": 9999,
  },
  "message": "Cannot find location or location's country",
  "name": "BadRequest",
  "originalError": undefined,
  "statusCode": 400,
}
`;

exports[`ClientsService create does not allow creating clients with the same cell number 1`] = `
Object {
  "extra": Object {
    "email": "walt@disney.com",
    "name": "Walt Disney",
    "phoneNumber": "+923214444444",
  },
  "message": "Client already exists",
  "name": "Conflict",
  "originalError": undefined,
  "statusCode": 409,
}
`;

exports[`ClientsService create does not allow creating clients with the same email address 1`] = `
Object {
  "extra": Object {
    "email": "cookies@monster.com",
    "name": "Walt Disney",
    "phoneNumber": "+923214444444",
  },
  "message": "Client already exists",
  "name": "Conflict",
  "originalError": undefined,
  "statusCode": 409,
}
`;

exports[`ClientsService create does not allow creating clients without having an owner 1`] = `
Object {
  "extra": Object {
    "ssoID": "4255c2ae-3f11-4a26-a401-3981c8845df3",
  },
  "message": "Current user has no owner",
  "name": "BadRequest",
  "originalError": undefined,
  "statusCode": 400,
}
`;

exports[`ClientsService find raises an error if an invalid number is specified 1`] = `
Object {
  "extra": Object {},
  "message": "ab is not a valid phone number",
  "name": "BadRequest",
  "originalError": undefined,
  "statusCode": 400,
}
`;
